libbeat/common/schema/mapstriface/mapstriface.go

```diff
@@ -70,26 +72,30 @@ type ConvMap struct {
 }
 
 // Map drills down in the data dictionary by using the key
-func (convMap ConvMap) Map(key string, event common.MapStr, data map[string]interface{}) *schema.Errors {
-	subData, ok := data[convMap.Key].(map[string]interface{})
+func (convMap ConvMap) Map(key string, event common.MapStr, data map[string]interface{}) multierror.Errors {
+	d, found := data[convMap.Key]
+	if !found {
+		err := schema.NewKeyNotFoundError(key)
+		err.Optional = convMap.Optional
+		return multierror.Errors{err}
+	}
+	subData, ok := d.(map[string]interface{})
 	if !ok {
-		err := schema.NewError(convMap.Key, "Error accessing sub-dictionary")
-		if convMap.Optional {
-			err.SetType(schema.OptionalType)
-		} else {
-			logp.Err("Error accessing sub-dictionary `%s`", convMap.Key)
-		}
-
-		errors := schema.NewErrors()
-		errors.AddError(err)
-
-		return errors
+		msg := fmt.Sprintf("expected dictionary, found %T", subData)
+		err := schema.NewWrongFormatError(convMap.Key, msg)
+		logp.Err(err.Error())
+		return multierror.Errors{err}
 	}
 
 	subEvent := common.MapStr{}
-	convMap.Schema.ApplyTo(subEvent, subData)
+	_, errors := convMap.Schema.ApplyTo(subEvent, subData)
+	for _, err := range errors {
+		if err, ok := err.(schema.KeyError); ok {
+			err.SetKey(convMap.Key + "." + err.Key())
+		}
```

Comment:
This is to have more meaningful errors on nested fields, previously a missing key would report something like `Key "segments" not found`, now it reports `key "total.segments" not found`.